gr-fec/lib/async_decoder_impl.cc

```diff
@@ -112,88 +107,56 @@ void async_decoder_impl::decode_unpacked(pmt::pmt_t msg)
+            d_decoder->generic_work(
+                (void*)&d_tmp_u8[i * d_decoder->get_input_size()],
+                (void*)&decoder_out[i * d_decoder->get_output_size()]);
+        }
     } else {
+        if (shift != 0.0) {
+            volk_32f_s32f_add_32f(d_tmp_f32.data(), f32in, shift, nbits_in);
+        } else {
+            memcpy(d_tmp_f32.data(), f32in, nbits_in * sizeof(float));
+        }
+
         for (size_t i = 0; i < nblocks; i++) {
-            d_decoder->generic_work((void*)&d_tmp_f32[i * d_decoder->get_input_size()],
-                                    (void*)&u8out[i * d_decoder->get_output_size()]);
+            d_decoder->generic_work(
+                (void*)&d_tmp_f32[i * d_decoder->get_input_size()],
+                (void*)&decoder_out[i * d_decoder->get_output_size()]);
         }
     }
 
+    // Note: when nblocks > 1, this only gets the iterations for the last
+    // decoder block.
     static const pmt::pmt_t iterations_key = pmt::mp("iterations");
     meta = pmt::dict_add(meta, iterations_key, pmt::mp(d_decoder->get_iterations()));
-    message_port_pub(d_out_port, pmt::cons(meta, outvec));
-}
-
-void async_decoder_impl::decode_packed(pmt::pmt_t msg)
-{
-    // extract input pdu
-    pmt::pmt_t meta(pmt::car(msg));
-    pmt::pmt_t bits(pmt::cdr(msg));
 
-    size_t o0 = 0;
-    size_t nbits_in = pmt::length(bits);
-    int nbits_out = nbits_in * d_decoder->rate();
-    int nbytes_out = nbits_out / 8;
-
-    // Check here if the frame size is larger than what we've
-    // allocated for in the constructor.
-    if (nbits_in > d_max_bits_in) {
-        throw std::runtime_error(
-            "async_decoder: Received frame larger than max frame size.");
-    }
-
-    d_decoder->set_frame_size(nbits_out);
-
-    pmt::pmt_t outvec(pmt::make_u8vector(nbytes_out, 0x00));
-    uint8_t* bytes_out = pmt::u8vector_writable_elements(outvec, o0);
-    const float* f32in = pmt::f32vector_elements(bits, o0);
-
-    if (strncmp(d_decoder->get_input_conversion(), "uchar", 5) == 0) {
-        volk_32f_s32f_multiply_32f(d_tmp_f32.data(), f32in, 48.0f, nbits_in);
-    } else {
-        memcpy(d_tmp_f32.data(), f32in, nbits_in * sizeof(float));
-    }
-
-    if (d_decoder->get_shift() != 0) {
-        for (size_t n = 0; n < nbits_in; n++)
-            d_tmp_f32[n] += d_decoder->get_shift();
-    }
-
-    if (strncmp(d_decoder->get_input_conversion(), "uchar", 5) == 0) {
-        // volk_32f_s32f_convert_8i(d_tmp_u8, d_tmp_f32, 1.0, nbits_in);
-        for (size_t n = 0; n < nbits_in; n++)
-            d_tmp_u8[n] = static_cast<uint8_t>(d_tmp_f32[n]);
-
-        d_decoder->generic_work((void*)d_tmp_u8.data(), (void*)d_bits_out.data());
-    } else {
-        d_decoder->generic_work((void*)d_tmp_f32.data(), (void*)d_bits_out.data());
+    if (d_packed) {
+        if (d_rev_pack)
+            d_pack.pack_rev(u8out, decoder_out, nitems_out);
+        else
+            d_pack.pack(u8out, decoder_out, nitems_out);
     }
```

Comment:
Would you mind to add `{..}` around the `if` and `else` branches? I know this is correct but it is also error prone. Wouldn't be the first time, I'd shoot myself in the foot with this.

Reply:
Done. I always write the braces. I think I didn't here because I was moving existing code around.